### ISSUE TITLE
We should not mark non-floating point Tensors as requirering gradients, ever

### ISSUE THREAD

**Issue by albanD** [pytorch/pytorch#37680]
We should not make non-floating point Tensor require gradients as it is not supported.
The fact that the backward function is an error is not enough as other ops following the faulty one are going to happily generate non-floating point Tensors that require gradients:

```
import torch
from torch import nn

t = torch.rand(10, requires_grad=True)
bad = torch.argmax(t)

res = bad + 2

res.sum().backward()
```

Faulty ops:
- argmax
- argmin ?

We should update the formulas to make sure the output does not require gradients but I am not even sure how argmax gets this not implemented node...

cc @ezyang @gchanan @zou3519 @SsnL @albanD @gqchen

**Comment by AdityaSoni19031997** [comment]
>We should not make non-floating point Tensor require gradients as it is not supported.

Just curious, Does this means that there's no grad defined for `integer` dtypes?

**Comment by albanD** [comment]
Gradients are only defined for continuous functions. So if you have an `integer` type, gradients don't really make sense.

**Comment by vadimkantorov** [comment]
If you are doing your own quantized type, it may have some sense, but only theoretically

**Comment by albanD** [comment]
Yes of course. But then you need to rewrite all the custom Functions.
And the argmax we provide has not gradient formula. So it should not return Tensors that require gradients.

**Comment by gchanan** [comment]
> If you are doing your own quantized type (with integer underlying type), it may have some sense

AFAIU "quantization-aware training" involves training with floating point dtypes with some fake quantization thrown in.  That doesn't mean it will always remain that way but we probably don't need to worry about this yet.

**Comment by vadimkantorov** [comment]
(I just wanted to point out that theoretically supporting `.grad` and backprop machinery / graph traversal may make sense for integer tensors, given of course custom implementation of all gradient computations. even if it's not what's happening now or in near term - this could be a way for users to do research on their own quantization schemes)

**Comment by tengerye** [comment]
Hi @albanD , may I ask if this issue has been solved yet? I find the following snippet does not work in current version, but works in earlier version, (e.g., 1.3.)

```
import torch
from torch import nn

t = torch.rand(10, requires_grad=True)
bad = torch.argmax(t)

res = bad + 2

**res.sum().float().backward()**
```

**Comment by albanD** [comment]
Hi,

This has been fixed in master. So only nightly builds for now. And it will be fixed in the upcoming 1.6 release.

**Comment by tengerye** [comment]
Hi, @albanD, thank you for your kind reply. May I ask does nightly support GPU? I follow the default official way to install it (`pip3 install --pre torch torchvision -f https://download.pytorch.org/whl/nightly/cpu/torch_nightly.html`), but only CPU version is installed.

**Comment by albanD** [comment]
You can follow the instructions on the get-started page to get the right command: https://pytorch.org/get-started/locally/
(The one you have is explicitly looking for cpu, make sure to select the cuda version you want in there).
